pkg/instance/service.go

```diff
@@ -265,11 +266,83 @@ func (s service) FindByNameAndGroup(instance string, group string) (*model.Insta
 	return s.instanceRepository.FindByNameAndGroup(instance, group)
 }
 
-func (s service) FindInstances(groups []*models.Group, presets bool) ([]*model.Instance, error) {
+type GroupWithInstances struct {
+	Name      string
+	Hostname  string
+	Instances []*model.Instance
+}
+
+func (s service) FindInstances(user *models.User, presets bool) ([]GroupWithInstances, error) {
+	groups := s.uniqueUserGroups(user)
+
 	groupNames := make([]string, len(groups))
 	for i, group := range groups {
 		groupNames[i] = group.Name
 	}
 
-	return s.instanceRepository.FindByGroupNames(groupNames, presets)
+	instances, err := s.instanceRepository.FindByGroupNames(groupNames, presets)
+	if err != nil {
+		return nil, err
+	}
+
+	return s.groupsWithInstances(instances), err
+}
+
+func (s service) uniqueUserGroups(user *models.User) []*models.Group {
+	groups := append(user.Groups, user.AdminGroups...)
+	return s.removeDuplicates(groups)
+}
+
+func (s service) groupsWithInstances(instances []*model.Instance) []GroupWithInstances {
+	groups := s.uniqueInstanceGroups(instances)
+	groupsWithInstances := make([]GroupWithInstances, len(groups))
+	for i, group := range groups {
+		groupsWithInstances[i].Name = group.Name
+		groupsWithInstances[i].Hostname = group.Hostname
+		groupsWithInstances[i].Instances = s.filterByGroupId(instances, func(instance *model.Instance) bool {
+			return instance.GroupName == group.Name
+		})
+	}
+	return groupsWithInstances
+}
+
+func (s service) uniqueInstanceGroups(instances []*model.Instance) []*models.Group {
+	groups := make([]*models.Group, len(instances))
+	for i, instance := range instances {
+		groups[i] = &models.Group{Name: instance.GroupName}
+	}
+	return s.removeDuplicates(groups)
+}
+
+func (s service) filterByGroupId(instances []*model.Instance, test func(instance *model.Instance) bool) (ret []*model.Instance) {
+	for _, instance := range instances {
+		if test(instance) {
+			ret = append(ret, instance)
+		}
+	}
+	return
 }
+
+func (s service) removeDuplicates(groups []*models.Group) []*models.Group {
```

Comment:
I don't think you are using the receiver 🤔 If so I would remove it. The receiver is just like another parameter. If you can remove it, please look up the call chain if you can remove it there too 😄

Reply:
My reason for define these as methods rather than package global functions is because I want their scope to be limited to the `service` struct.